# 이분 탐색
- 이분 탐색이란, 오름차순으로 정렬된 배열을 반복적으로 반으로 나누어 target이 선택될 때까지 탐색하는 알고리즘이다.
- 시간복잡도 : O(logN)
- 반복문과 재귀 두 가지 방법을 사용할 수 있음.
- 자료의 중간값(mid)이 찾고자 하는 값(target)인지 비교함.

# target과 mid 값이 같을 때까지
- (1) target < mid 일 때, end = mid (절반의 왼쪽 탐색)
- (2) target > mid 일 때, start = mid (절반의 오른쪽 탐색)

In [ ]:
# 01
# 이분 탐색
def binary_search(start, end, target):
  if start > end:
    return 0

  mid_index = (start + end) // 2
  mid = arr[mid_index]

  # 값을 찾았을 때
  if target == mid:
    return 1
  # 왼쪽 탐색
  elif target < mid:
    return binary_search(start, mid_index - 1, target)
  # 오른쪽 탐색
  else:
    return binary_search(mid_index + 1, end, target)

N = int(input())
arr = list(map(int, input().split()))
arr.sort()

M = int(input())
target_list = list(map(int, input().split()))

for target in target_list:
  print(binary_search(0, N - 1, target))

5
4 1 5 2 3
5
1 3 7 9 5
1
1
0
0
1


In [ ]:
# 02 - 시간 초과
def binary_search(start, end, target):
  if start > end:
    return 0

  mid_index = (start + end) // 2
  mid = arr[mid_index]

  # 조건문의 순서가 의미있음. mid == target이 무조건 맨 아래에 있어야함.
  if mid > target:
    return binary_search(start, mid_index - 1, target)
  elif mid < target:
    return binary_search(mid_index + 1, end, target)
  else:     # mid == target
    # return 형태 기억하기
    return 1 + binary_search(start, mid_index - 1, target) + binary_search(mid_index + 1, end, target)


N = int(input())
arr = list(map(int, input().split()))
arr.sort()

M = int(input())
target_list = list(map(int, input().split()))

result = []
for target in target_list:
  result.append(binary_search(0, N - 1, target))

print(*result)

10
6 3 2 10 10 10 -10 -10 7 3
8
10 9 -5 2 3 4 5 -10
3 0 0 1 2 0 0 2


# 아이디어
- 위 1번 문제와 달리 이 문제는 count 문제이므로 이분 탐색 시 주의해야 할 점이 있다.
- 예를 들어 오름차순 배열 시, [6 8 10 10 10 10 10] 이라는 배열이 되고 이는 mid 값이 10이며 만약 target 값이 10일 때, mid를 기준으로 좌우 모두에 10이 존재하게 된다. 따라서 mid = target일 때, 좌우 모두 고려해주어야 할 것이다.

In [ ]:
# 02 - 다시 풀어보기(이분 탐색 x, 딕셔너리 사용)

N = int(input())
arr = sorted(list(map(int, input().split())))
M = int(input())
targets = list(map(int, input().split()))

# dic에 입력 요소들의 개수 저장
dic = {}
for x in arr:
  if x in dic:
    dic[x] += 1
  else:
    dic[x] = 1

# key값을 비교하며 확인
for target in targets:
  if target in dic:
    print(dic[target], end = ' ')
  else:
    print(0, end = ' ')

10
6 3 2 10 10 10 -10 -10 7 3
8
10 9 -5 2 3 4 5 -10
3 0 0 1 2 0 0 2 

# 문제 풀면서 힘들었던 점
- return 1 + binary_search(start, mid_index - 1, target) + binary_search(mid_index + 1, end, target)의 형태를 생각하지 못했다. +1 처리가 어려웠음.
- 이분 탐색(이진 탐색) 문제여서 다른 방법을 고려하지 않았음. 이는 딕셔너리로 푸는 것이 훨씬 좋은 문제임.

In [ ]:
# 03
def parametric_search(start, end):
  global answer     # 최대 길이를 저장할 변수
  if start > end:
    return          # 불필요한 값 반환 x

  mid = (start + end) // 2
  cnt = sum(lan // mid for lan in lans)

  if N <= cnt:
    answer = mid
    parametric_search(mid + 1, end)
  else:
    parametric_search(start, mid - 1)

K, N = map(int, input().split())
lans = sorted([int(input()) for _ in range(K)])

answer = 0
parametric_search(1, lans[-1])
print(answer)

4 11
802
743
457
539
200


# Parametric Search
- 원래 주어진 문제를 결정문제로 변형하여 이분탐색을 통해 해결하는 것
- ex) "고기 200g을 잘라라" => "지금 자른 고기가 200g보다 가벼운가/무거운가"

# parametric search를 활용하기 위한 조건
- (1) 특정 조건을 만족하는 최댓값/최솟값을 구하는 형식의 문제여야 함. 혹은 이러한 형태로 변형이 가능해야 함. ex) "200g 이하의 고기 중 최대 무게를 구하여라"
- (2) 최댓값을 구하는 문제의 경우, 어떤 값이 조건을 만족하면 그 값보다 작은 값은 모두 조건을 만족해야 한다. (최솟값의 경우엔 그 값보다 큰 값은 모두 조건을 만족해야 한다.)
- (3) 답의 범위가 이산적이거나 허용 오차 범위가 있어야 함.

# 아이디어
- 우선 제일 작은 수를 기준으로 모든 값을 나눈 몫을 계산해본다.
- (1) N > 모든 몫의 합
- ex1) 457 // 457 = 1, 539 // 457 = 1, 743 // 457 = 1, 802 // 457 = 1 => 1 + 1 + 1 + 1 = "4"
- N = 11 > 4 이므로 457을 반으로 나눈 값(사실상 0과 457의 mid 값)으로 다시 모든 값들을 나눈 몫을 계산. 이분 탐색을 통해 계속 안으로 파고들어가 최대 길이를 구한다.
- (2) N < 모든 몫의 합
- ex2) 2 // 2 = 1, 4 // 2 = 2, 5 // 2 = 2, 6 // 2 = 3 => 1 + 2 + 2 + 3 = "8"
- N가 4라고 예를 들어보자. N = 4 < 8 이므로 제일 작은 값인 2와 제일 큰 값인 6의 mid 값으로 다시 모든 값들을 나눈 몫을 계산. 이분 탐색을 통해 계속 안으로 파고들어가 최대 길이를 구한다.
- (3) N = 모든 몫의 합 => (2)번과 같은 방법으로 고려하면 될듯

# 위의 아이디어와 조금 다르게 풀었음.

In [ ]:
# 04
def parametric_search(start, end):
  global answer
  if start > end:
    return

  mid = (start + end) // 2
  sum_len = sum((tree - mid) for tree in trees if tree > mid)

  if sum_len >= M:
    answer = mid
    parametric_search(mid + 1, end)
  else:
    parametric_search(start, mid - 1)

N, M = map(int, input().split())
trees = sorted(list(map(int, input().split())))

answer = 0
parametric_search(0, trees[-1])
print(answer)

5 20
4 42 40 26 46
36


# 아이디어
- 나무 높이 리스트를 sorted를 통해 오름차순 정렬을 해줌.
- 이번에도 parametric search를 이용할 것이다. 제일 큰 수와 0의 mid값 기준으로 이분 탐색을 시작한다.
- tree - mid < 0일 때를 고려해야함.

In [ ]:
# 05
def binary_search(start, end):
  global answer
  while start <= end:
    mid = (start + end) // 2
    current = wifi[0]
    count = 1

    for i in range(1, len(wifi)):
      if wifi[i] >= current + mid:      # current 값이 계속 변하는 것을 인지해야함.
        count += 1
        current = wifi[i]

    if count >= C:
      start = mid + 1
      answer = mid
    else:
      end = mid - 1

N, C = map(int, input().split())
wifi = sorted(int(input()) for _ in range(N))     # list()를 추가로 사용할 필요 X, sorted 자체가 리스트를 반환하기 때문.

answer = 0
binary_search(1, wifi[-1] - wifi[0])
# start와 end 둘 다 집과 집 사이의 거리를 나타내므로 start를 1로 시작하는 것임.
print(answer)

5 3
1
2
8
4
9
3


# 아이디어
- 위의 3,4번 문제처럼 값으로 접근하는 것이 아닌, 인덱스로 접근하여 이분 탐색을 응용하는 문제일듯? (X)
- 이 문제에서 이진 탐색을 적용하는 방식은 "가능한 공유기 간 거리"를 기준으로 탐색하는 것이다.
- (1) mid를 현재 가능한 거리의 중간값으로 설정
- (2) 이 거리(mid)를 유지하면서 공유기를 최대한 설치할 수 있는지 확인
- (3) 공유기를 원하는 개수(C)만큼 설치할 수 있으면, 더 멀리 배치할 수 있는지 확인 (start = mid + 1)
- (4) 불가능하면 거리를 줄임 (end = mid - 1)

- 이 문제는 재귀 대신 함수 안에 while 반복문을 사용하여 처리를 한다.

- <코드의 결함> : (148)의 경우 계산되지만, (149)의 경우 계산되지 않는다.

In [ ]:
# 06

N = int(input())    # 행렬의 크기
k = int(input())    # 찾고 싶은 k번째 수
start = 1
end = k   # 최댓값을 k로 설정 (이진 탐색 범위)

while start <= end:
  mid = (start + end) // 2    # 중간값 (이 값이 k번째 수인지 확인할 후보)
  cnt = 0   # mid 이하의 숫자 개수 세기

  for i in range(1, N + 1):
    cnt += min(mid // i, N)   # mid // i => i번째 행에서 mid 이하인 원소의 개수
                              # 한 행에는 원소가 N개를 넘을 수 없으므로 N과 비교하여 최솟값을 cnt에 넣어줘야함.

  if cnt >= k:    # 만약 특정 수 X가 행렬의 k번째 수보다 크거나 같다면
    result = mid
    end = mid - 1
  else:
    start = mid + 1

print(result)

# N = 3 (배열 A와 B의 인덱스는 1부터 시작한다는 것을 인지)
- 1*1 = 1 / 1*2 = 2 / 1*3 = 3
- 2*1 = 2 / 2*2 = 4 / 2*3 = 6
- 3*1 = 3 / 3*2 = 6 / 3*3 = 9

- B = [1, 2, 3, 2, 4, 6, 3, 6, 9]
- 오름차순 정렬을 한 B = [1, 2, 2, 3, 3, 4, 6, 6, 9]

- k = 7, "6" return

# 아이디어 (K번째 작은 수 찾기)
- ex) N = 20일 때, N*N에서 20보다 작거나 같은 수는
- 1행 기준 : 1 ~ 20 => 20 // 1 개
- 2행 기준 : 2, 4, 6, 8... => 20 // 2 개
- 3행 기준 : 3, 6, 9, 12... => 20 // 3 개
- 4행 기준 : 4, 8, 12, 16... => 20 // 4 개
- N행 기준 : N, N*2, N*3... => 20 // N 개
- 이 중 찾는값 // i 가 N보다 큰 경우가 존재하므로, 최대 N까지만 갖게 설정

# 핵심 발상
- 단순히 리스트를 만들어 정렬하면 비효율적임. 대신에 특정 숫자 X가 몇 번째로 작은 수인지 계산하는 방식으로 접근하면 된다. 즉, X 이하의 숫자가 몇 개 있는지를 확인하는 방식으로 이진 탐색을 진행.

In [2]:
# 07

def binary_search(num):
  start, end = 0, len(arr) - 1
  while start <= end:
    mid = (start + end) // 2
    if arr[mid] < num:
      start = mid + 1
    else:
      end = mid - 1
  return start

N = int(input())
A = list(map(int, input().split()))
arr = [0]

for i in range(N):
  if arr[-1] < A[i]:
    arr.append(A[i])
  elif arr[-1] >= A[i]:
    idx = binary_search(A[i])
    arr[idx] = A[i]

print(len(arr) - 1)

6
10 20 10 30 20 50
4
